### ISSUE TITLE
scraper being killed after blowing its memory limit

### ISSUE THREAD

**Issue by otherchirps** [otherchirps/nsw_gov_docs#1]
# Scraper error

Scraper appears to be killed immediately after requirement installation, with the message:

> /start: line 24:    21 Killed                  setuidgid scraper $(eval echo ${command})

This "start" script is not part of this scraper's codebase.  

Also, the scraper itself wasn't altered before this started happening.  So perhaps the morph framework has altered lately, and as a side effect, this scraper is no longer playing as nicely as it should?  

Or maybe it's starting to hit a memory limit?  Don't know yet.
# Oddness

Something seems to run at some point though, since the morph run status log shows things like:

```
Manually ran revision 97c8157a and failed 2015-06-01.
run time 213.1 s
533 records added, 533 records removed in the database
11 pages scraped 
```

The number of records added appears a bit random at the moment, but if it was being killed before running, you'd think this would be zero.
# Install logs

Here is the output of the scraper install & attempt to run.  Might need to add more debug logs to see if it actually does get to run at all or not:

```
Injecting configuration and compiling...
[1G-----> Python app detected
[1G-----> Stack changed, re-installing runtime
[1G-----> Installing runtime (python-2.7.6)
[1G-----> Installing dependencies with pip
[1G       Collecting scraperwiki==0.4.1 (from -r requirements.txt (line 1))
[1G       Downloading scraperwiki-0.4.1.tar.gz
[1G       Collecting Scrapy==0.24.6 (from -r requirements.txt (line 2))
[1G       Downloading Scrapy-0.24.6-py2-none-any.whl (444kB)
[1G       Collecting service-identity==14.0.0 (from -r requirements.txt (line 3))
[1G       Downloading service_identity-14.0.0-py2.py3-none-any.whl
[1G       Collecting requests (from scraperwiki==0.4.1->-r requirements.txt (line 1))
[1G       Downloading requests-2.7.0-py2.py3-none-any.whl (470kB)
[1G       Collecting sqlalchemy (from scraperwiki==0.4.1->-r requirements.txt (line 1))
[1G       Downloading SQLAlchemy-1.0.4.tar.gz (4.6MB)
[1G       Collecting alembic (from scraperwiki==0.4.1->-r requirements.txt (line 1))
[1G       Downloading alembic-0.7.6.tar.gz (623kB)
[1G       Collecting cssselect>=0.9 (from Scrapy==0.24.6->-r requirements.txt (line 2))
[1G       Downloading cssselect-0.9.1.tar.gz
[1G       Collecting queuelib (from Scrapy==0.24.6->-r requirements.txt (line 2))
[1G       Downloading queuelib-1.2.2-py2.py3-none-any.whl
[1G       Collecting pyOpenSSL (from Scrapy==0.24.6->-r requirements.txt (line 2))
[1G       Downloading pyOpenSSL-0.15.1-py2.py3-none-any.whl (102kB)
[1G       Collecting w3lib>=1.8.0 (from Scrapy==0.24.6->-r requirements.txt (line 2))
[1G       Downloading w3lib-1.11.0-py2.py3-none-any.whl
[1G       Collecting lxml (from Scrapy==0.24.6->-r requirements.txt (line 2))
[1G       Downloading lxml-3.4.4.tar.gz (3.5MB)
[1G       Building lxml version 3.4.4.
[1G       Building without Cython.
[1G       Using build configuration of libxslt 1.1.28
[1G       /app/.heroku/python/lib/python2.7/distutils/dist.py:267: UserWarning: Unknown distribution option: 'bugtrack_url'
[1G       warnings.warn(msg)
[1G       Collecting Twisted>=10.0.0 (from Scrapy==0.24.6->-r requirements.txt (line 2))
[1G       Downloading Twisted-15.2.1.tar.bz2 (4.6MB)
[1G       Collecting six>=1.5.2 (from Scrapy==0.24.6->-r requirements.txt (line 2))
[1G       Downloading six-1.9.0-py2.py3-none-any.whl
[1G       Collecting characteristic>=14.0.0 (from service-identity==14.0.0->-r requirements.txt (line 3))
[1G       Downloading characteristic-14.3.0-py2.py3-none-any.whl
[1G       Collecting pyasn1-modules (from service-identity==14.0.0->-r requirements.txt (line 3))
[1G       Downloading pyasn1-modules-0.0.5.tar.gz
[1G       Collecting pyasn1 (from service-identity==14.0.0->-r requirements.txt (line 3))
[1G       Downloading pyasn1-0.1.7.tar.gz (68kB)
[1G       Collecting Mako (from alembic->scraperwiki==0.4.1->-r requirements.txt (line 1))
[1G       Downloading Mako-1.0.1.tar.gz (473kB)
[1G       Collecting cryptography>=0.7 (from pyOpenSSL->Scrapy==0.24.6->-r requirements.txt (line 2))
[1G       Downloading cryptography-0.9.tar.gz (302kB)
[1G       Collecting zope.interface>=3.6.0 (from Twisted>=10.0.0->Scrapy==0.24.6->-r requirements.txt (line 2))
[1G       Downloading zope.interface-4.1.2.tar.gz (919kB)
[1G       Collecting MarkupSafe>=0.9.2 (from Mako->alembic->scraperwiki==0.4.1->-r requirements.txt (line 1))
[1G       Downloading MarkupSafe-0.23.tar.gz
[1G       Collecting idna (from cryptography>=0.7->pyOpenSSL->Scrapy==0.24.6->-r requirements.txt (line 2))
[1G       Downloading idna-2.0-py2.py3-none-any.whl (61kB)
[1G       Collecting enum34 (from cryptography>=0.7->pyOpenSSL->Scrapy==0.24.6->-r requirements.txt (line 2))
[1G       Downloading enum34-1.0.4.tar.gz
[1G       Collecting ipaddress (from cryptography>=0.7->pyOpenSSL->Scrapy==0.24.6->-r requirements.txt (line 2))
[1G       Downloading ipaddress-1.0.7-py27-none-any.whl
[1G       Collecting cffi>=0.8 (from cryptography>=0.7->pyOpenSSL->Scrapy==0.24.6->-r requirements.txt (line 2))
[1G       Downloading cffi-1.1.0.tar.gz (323kB)
[1G       Collecting pycparser (from cffi>=0.8->cryptography>=0.7->pyOpenSSL->Scrapy==0.24.6->-r requirements.txt (line 2))
[1G       Downloading pycparser-2.13.tar.gz (299kB)
[1G       Installing collected packages: pycparser, cffi, ipaddress, enum34, idna, MarkupSafe, zope.interface, cryptography, Mako, pyasn1, pyasn1-modules, characteristic, six, Twisted, lxml, w3lib, pyOpenSSL, queuelib, cssselect, alembic, sqlalchemy, requests, service-identity, Scrapy, scraperwiki
[1G       Running setup.py install for pycparser
[1G       Build the lexing/parsing tables
[1G       Running setup.py install for cffi
[1G       building '_cffi_backend' extension
[1G       gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -DUSE__THREAD -I/app/.heroku/python/include/python2.7 -c c/_cffi_backend.c -o build/temp.linux-x86_64-2.7/c/_cffi_backend.o
[1G       gcc -pthread -shared build/temp.linux-x86_64-2.7/c/_cffi_backend.o -L/app/.heroku/python/lib -lffi -lpython2.7 -o build/lib.linux-x86_64-2.7/_cffi_backend.so
[1G       
[1G       Running setup.py install for enum34
[1G       
[1G       Running setup.py install for MarkupSafe
[1G       building 'markupsafe._speedups' extension
[1G       gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/app/.heroku/python/include/python2.7 -c markupsafe/_speedups.c -o build/temp.linux-x86_64-2.7/markupsafe/_speedups.o
[1G       gcc -pthread -shared build/temp.linux-x86_64-2.7/markupsafe/_speedups.o -L/app/.heroku/python/lib -lpython2.7 -o build/lib.linux-x86_64-2.7/markupsafe/_speedups.so
[1G       Running setup.py install for zope.interface
[1G       building 'zope.interface._zope_interface_coptimizations' extension
[1G       gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/app/.heroku/python/include/python2.7 -c src/zope/interface/_zope_interface_coptimizations.c -o build/temp.linux-x86_64-2.7/src/zope/interface/_zope_interface_coptimizations.o
[1G       gcc -pthread -shared build/temp.linux-x86_64-2.7/src/zope/interface/_zope_interface_coptimizations.o -L/app/.heroku/python/lib -lpython2.7 -o build/lib.linux-x86_64-2.7/zope/interface/_zope_interface_coptimizations.so
[1G       Skipping installation of /app/.heroku/python/lib/python2.7/site-packages/zope/__init__.py (namespace package)
[1G       Installing /app/.heroku/python/lib/python2.7/site-packages/zope.interface-4.1.2-py2.7-nspkg.pth
[1G       Running setup.py install for cryptography
[1G       zip_safe flag not set; analyzing archive contents...
[1G       six: module references __path__
[1G       Installed /tmp/pip-build-VOb276/cryptography/.eggs/six-1.9.0-py2.7.egg
[1G       Searching for pyasn1
[1G       Reading https://pypi.python.org/simple/pyasn1/
[1G       Best match: pyasn1 0.1.7
[1G       Downloading https://pypi.python.org/packages/2.7/p/pyasn1/pyasn1-0.1.7-py2.7.egg#md5=15f079cabee01402bf86ca8b83356469
[1G       Processing pyasn1-0.1.7-py2.7.egg
[1G       Moving pyasn1-0.1.7-py2.7.egg to /tmp/pip-build-VOb276/cryptography/.eggs
[1G       Installed /tmp/pip-build-VOb276/cryptography/.eggs/pyasn1-0.1.7-py2.7.egg
[1G       building '_Cryptography_cffi_e7d09016xc302a38b' extension
[1G       gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/app/.heroku/python/include/python2.7 -c src/cryptography/hazmat/bindings/__pycache__/_Cryptography_cffi_e7d09016xc302a38b.c -o build/temp.linux-x86_64-2.7/src/cryptography/hazmat/bindings/__pycache__/_Cryptography_cffi_e7d09016xc302a38b.o
[1G       gcc -pthread -shared build/temp.linux-x86_64-2.7/src/cryptography/hazmat/bindings/__pycache__/_Cryptography_cffi_e7d09016xc302a38b.o -L/app/.heroku/python/lib -lssl -lcrypto -lpython2.7 -o build/lib.linux-x86_64-2.7/cryptography/_Cryptography_cffi_e7d09016xc302a38b.so
[1G       building '_Cryptography_cffi_590da19fxffc7b1ce' extension
[1G       gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/app/.heroku/python/include/python2.7 -c src/cryptography/hazmat/bindings/__pycache__/_Cryptography_cffi_590da19fxffc7b1ce.c -o build/temp.linux-x86_64-2.7/src/cryptography/hazmat/bindings/__pycache__/_Cryptography_cffi_590da19fxffc7b1ce.o
[1G       gcc -pthread -shared build/temp.linux-x86_64-2.7/src/cryptography/hazmat/bindings/__pycache__/_Cryptography_cffi_590da19fxffc7b1ce.o -L/app/.heroku/python/lib -lpython2.7 -o build/lib.linux-x86_64-2.7/cryptography/_Cryptography_cffi_590da19fxffc7b1ce.so
[1G       building '_Cryptography_cffi_26cb75b8x62b488b1' extension
[1G       gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/app/.heroku/python/include/python2.7 -c src/cryptography/hazmat/bindings/__pycache__/_Cryptography_cffi_26cb75b8x62b488b1.c -o build/temp.linux-x86_64-2.7/src/cryptography/hazmat/bindings/__pycache__/_Cryptography_cffi_26cb75b8x62b488b1.o
[1G       gcc -pthread -shared build/temp.linux-x86_64-2.7/src/cryptography/hazmat/bindings/__pycache__/_Cryptography_cffi_26cb75b8x62b488b1.o -L/app/.heroku/python/lib -lpython2.7 -o build/lib.linux-x86_64-2.7/cryptography/_Cryptography_cffi_26cb75b8x62b488b1.so
[1G       Running setup.py install for Mako
[1G       Installing mako-render script to /app/.heroku/python/bin
[1G       Running setup.py install for pyasn1
[1G       Running setup.py install for pyasn1-modules
[1G       
[1G       
[1G       Running setup.py install for Twisted
[1G       gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/app/.heroku/python/include/python2.7 -c conftest.c -o conftest.o
[1G       building 'twisted.runner.portmap' extension
[1G       gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/app/.heroku/python/include/python2.7 -c twisted/runner/portmap.c -o build/temp.linux-x86_64-2.7/twisted/runner/portmap.o
[1G       gcc -pthread -shared build/temp.linux-x86_64-2.7/twisted/runner/portmap.o -L/app/.heroku/python/lib -lpython2.7 -o build/lib.linux-x86_64-2.7/twisted/runner/portmap.so
[1G       building 'twisted.test.raiser' extension
[1G       gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/app/.heroku/python/include/python2.7 -c twisted/test/raiser.c -o build/temp.linux-x86_64-2.7/twisted/test/raiser.o
[1G       gcc -pthread -shared build/temp.linux-x86_64-2.7/twisted/test/raiser.o -L/app/.heroku/python/lib -lpython2.7 -o build/lib.linux-x86_64-2.7/twisted/test/raiser.so
[1G       building 'twisted.python.sendmsg' extension
[1G       gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/app/.heroku/python/include/python2.7 -c twisted/python/sendmsg.c -o build/temp.linux-x86_64-2.7/twisted/python/sendmsg.o
[1G       gcc -pthread -shared build/temp.linux-x86_64-2.7/twisted/python/sendmsg.o -L/app/.heroku/python/lib -lpython2.7 -o build/lib.linux-x86_64-2.7/twisted/python/sendmsg.so
[1G       changing mode of build/scripts-2.7/pyhtmlizer from 644 to 755
[1G       changing mode of build/scripts-2.7/tap2rpm from 644 to 755
[1G       changing mode of build/scripts-2.7/manhole from 644 to 755
[1G       changing mode of build/scripts-2.7/tap2deb from 644 to 755
[1G       changing mode of build/scripts-2.7/trial from 644 to 755
[1G       changing mode of build/scripts-2.7/twistd from 644 to 755
[1G       changing mode of build/scripts-2.7/tkconch from 644 to 755
[1G       changing mode of build/scripts-2.7/conch from 644 to 755
[1G       changing mode of build/scripts-2.7/ckeygen from 644 to 755
[1G       changing mode of build/scripts-2.7/cftp from 644 to 755
[1G       changing mode of build/scripts-2.7/mailmail from 644 to 755
[1G       changing mode of build/scripts-2.7/lore from 644 to 755
[1G       changing mode of /app/.heroku/python/bin/tkconch to 755
[1G       changing mode of /app/.heroku/python/bin/conch to 755
[1G       changing mode of /app/.heroku/python/bin/pyhtmlizer to 755
[1G       changing mode of /app/.heroku/python/bin/tap2rpm to 755
[1G       changing mode of /app/.heroku/python/bin/manhole to 755
[1G       changing mode of /app/.heroku/python/bin/ckeygen to 755
[1G       changing mode of /app/.heroku/python/bin/tap2deb to 755
[1G       changing mode of /app/.heroku/python/bin/mailmail to 755
[1G       changing mode of /app/.heroku/python/bin/trial to 755
[1G       changing mode of /app/.heroku/python/bin/twistd to 755
[1G       changing mode of /app/.heroku/python/bin/lore to 755
[1G       changing mode of /app/.heroku/python/bin/cftp to 755
[1G       Running setup.py install for lxml
[1G       Building lxml version 3.4.4.
[1G       Building without Cython.
[1G       Using build configuration of libxslt 1.1.28
[1G       /app/.heroku/python/lib/python2.7/distutils/dist.py:267: UserWarning: Unknown distribution option: 'bugtrack_url'
[1G       warnings.warn(msg)
[1G       building 'lxml.etree' extension
[1G       gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/include/libxml2 -I/tmp/pip-build-VOb276/lxml/src/lxml/includes -I/app/.heroku/python/include/python2.7 -c src/lxml/lxml.etree.c -o build/temp.linux-x86_64-2.7/src/lxml/lxml.etree.o -w
[1G       gcc -pthread -shared build/temp.linux-x86_64-2.7/src/lxml/lxml.etree.o -L/app/.heroku/python/lib -lxslt -lexslt -lxml2 -lz -lm -lpython2.7 -o build/lib.linux-x86_64-2.7/lxml/etree.so
[1G       building 'lxml.objectify' extension
[1G       gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/include/libxml2 -I/tmp/pip-build-VOb276/lxml/src/lxml/includes -I/app/.heroku/python/include/python2.7 -c src/lxml/lxml.objectify.c -o build/temp.linux-x86_64-2.7/src/lxml/lxml.objectify.o -w
[1G       gcc -pthread -shared build/temp.linux-x86_64-2.7/src/lxml/lxml.objectify.o -L/app/.heroku/python/lib -lxslt -lexslt -lxml2 -lz -lm -lpython2.7 -o build/lib.linux-x86_64-2.7/lxml/objectify.so
[1G       
[1G       
[1G       
[1G       Running setup.py install for cssselect
[1G       Running setup.py install for alembic
[1G       Installing alembic script to /app/.heroku/python/bin
[1G       Running setup.py install for sqlalchemy
[1G       building 'sqlalchemy.cprocessors' extension
[1G       gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/app/.heroku/python/include/python2.7 -c lib/sqlalchemy/cextension/processors.c -o build/temp.linux-x86_64-2.7/lib/sqlalchemy/cextension/processors.o
[1G       gcc -pthread -shared build/temp.linux-x86_64-2.7/lib/sqlalchemy/cextension/processors.o -L/app/.heroku/python/lib -lpython2.7 -o build/lib.linux-x86_64-2.7/sqlalchemy/cprocessors.so
[1G       building 'sqlalchemy.cresultproxy' extension
[1G       gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/app/.heroku/python/include/python2.7 -c lib/sqlalchemy/cextension/resultproxy.c -o build/temp.linux-x86_64-2.7/lib/sqlalchemy/cextension/resultproxy.o
[1G       gcc -pthread -shared build/temp.linux-x86_64-2.7/lib/sqlalchemy/cextension/resultproxy.o -L/app/.heroku/python/lib -lpython2.7 -o build/lib.linux-x86_64-2.7/sqlalchemy/cresultproxy.so
[1G       building 'sqlalchemy.cutils' extension
[1G       gcc -pthread -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/app/.heroku/python/include/python2.7 -c lib/sqlalchemy/cextension/utils.c -o build/temp.linux-x86_64-2.7/lib/sqlalchemy/cextension/utils.o
[1G       gcc -pthread -shared build/temp.linux-x86_64-2.7/lib/sqlalchemy/cextension/utils.o -L/app/.heroku/python/lib -lpython2.7 -o build/lib.linux-x86_64-2.7/sqlalchemy/cutils.so
[1G       
[1G       
[1G       
[1G       Running setup.py install for scraperwiki
[1G       Successfully installed Mako-1.0.1 MarkupSafe-0.23 Scrapy-0.24.6 Twisted-15.2.1 alembic-0.7.6 cffi-1.1.0 characteristic-14.3.0 cryptography-0.9 cssselect-0.9.1 enum34-1.0.4 idna-2.0 ipaddress-1.0.7 lxml-3.4.4 pyOpenSSL-0.15.1 pyasn1-0.1.7 pyasn1-modules-0.0.5 pycparser-2.13 queuelib-1.2.2 requests-2.7.0 scraperwiki-0.4.1 service-identity-14.0.0 six-1.9.0 sqlalchemy-1.0.4 w3lib-1.11.0 zope.interface-4.1.2
[1G       
[1G-----> Discovering process types
[1G       Procfile declares types -> scraper
Injecting scraper and running...
/start: line 24:    21 Killed                  setuidgid scraper $(eval echo ${command})
```

**Comment by otherchirps** [comment]
scraper runs & completes without error, when running manually locally.  So it must be fighting the morph environment for some reason.

**Comment by otherchirps** [comment]
Attempted to run a local instance of morph (following the [local instance docs](https://github.com/openaustralia/morph#production-devops-development) ), but "vagrant up local" dies with:

```
TASK: [docker-server-linode | Ensure that the docker package is installed] **** 
failed: [local] => {"failed": true}
msg: No package matching 'lxc-docker' is available

FATAL: all hosts have already failed -- aborting

PLAY RECAP ******************************************************************** 
           to retry, use: --limit @/home/chris/playbook.retry

local                      : ok=12   changed=0    unreachable=0    failed=1 
```

Also tried running with [morph-cli](https://github.com/openaustralia/morph-cli), as suggested in the [docs](https://morph.io/documentation), in "Testing your scraper locally", but it seems that installing the libraries takes too long to complete the test.  It exits before finishing with:

```
/var/lib/gems/2.1.0/gems/rest-client-1.8.0/lib/restclient/request.rb:427:in `rescue in transmit': Request Timeout (RestClient::RequestTimeout)
    from /var/lib/gems/2.1.0/gems/rest-client-1.8.0/lib/restclient/request.rb:350:in `transmit'
    from /var/lib/gems/2.1.0/gems/rest-client-1.8.0/lib/restclient/request.rb:176:in `execute'
    from /var/lib/gems/2.1.0/gems/rest-client-1.8.0/lib/restclient/request.rb:41:in `execute'
    from /var/lib/gems/2.1.0/gems/morph-cli-0.2.2/lib/morph-cli.rb:26:in `execute'
    from /var/lib/gems/2.1.0/gems/morph-cli-0.2.2/bin/morph:32:in `execute'
    from /var/lib/gems/2.1.0/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
    from /var/lib/gems/2.1.0/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
    from /var/lib/gems/2.1.0/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
    from /var/lib/gems/2.1.0/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
    from /var/lib/gems/2.1.0/gems/morph-cli-0.2.2/bin/morph:63:in `<top (required)>'
    from /usr/local/bin/morph:23:in `load'
    from /usr/local/bin/morph:23:in `<main>'
```

Might have to log on the "live" version, if can't test locally.

**Comment by otherchirps** [comment]
Ok. Disabled the morph-cli restclient request timeout, as per [stackoverflow](http://stackoverflow.com/a/5445421/75188).  Same error happens, so can now reproduce the scraper's problem "locally".

Altered the morph-cli file perfoming the request call: 
/var/lib/gems/2.1.0/gems/morph-cli-0.2.2/lib/morph-cli.rb

~~Dodgy. todo: submit a patch for review, for upstream morph-cli.~~

Submitted pull request: https://github.com/openaustralia/morph-cli/pull/9

**Comment by otherchirps** [comment]
Cool. It seems the scraper does get to start.  With more logs:

```
Uploading and running...
Injecting configuration and compiling...
Injecting scraper and running...
Scraper running...

spawned spider...
loaded settings...
starting spider crawl...
/start: line 24:    21 Killed                  setuidgid scraper $(eval echo ${command})
```

This shows the script starts, at least.

**Comment by otherchirps** [comment]
Script is being killed externally.  Exception handler not getting a bite of the cherry. 

The scraper does fetch data for a little while before being killed.
That's the "Killed" output line.  The echo written after the kill executes.

Implies it's hitting a resource limit, I guess.  Memory leak?

**Comment by otherchirps** [comment]
Looks like a leak.  But the scraper's RAM usage seems high to start with.
Starts at a baseline of reportedly 79MB.  This hits around 100 before being killed.  Last reported usage:

```
[scrapy] INFO: Memory usage: 100768 (kb)
```

A full successful run from my local machine tops out with:

```
[scrapy] INFO: Memory usage: 269528 (kb)
```

Yikes. Need to trim this.

It seems the scrapy docs have some [advice on this](http://doc.scrapy.org/en/latest/faq.html#my-scrapy-crawler-has-memory-leaks-what-can-i-do)... 

Hm. [Not encouraging though](http://doc.scrapy.org/en/latest/topics/leaks.html#leaks-without-leaks):

> Sometimes, you may notice that the memory usage of your Scrapy process will only increase, but never decrease. Unfortunately, this could happen even though neither Scrapy nor your project are leaking memory. This is due to a (not so well) known problem of Python, which may not return released memory to the operating system in some cases.

**Comment by otherchirps** [comment]
Enabled scrapy's [memory debugger](http://doc.scrapy.org/en/latest/topics/extensions.html#module-scrapy.contrib.memdebug): 

```
'item_scraped_count': 37547,
'memdebug/gc_garbage_count': 0,
'memdebug/live_refs/LegislativeAssemblyTabledDocsSpider': 1,
'memusage/max': 298221568,
'memusage/startup': 73551872,
```

The zero garbage_count, and single live_ref show there aren't any leaks caused by the code...
So chances are it's the runtime holding previously allocated memory in a freelist, or something, and not yielding it back to the OS.

One option might be to split the processing.  Currently the crawler walks each parliament session page, then walks all links on those pages.  Perhaps we can have a subprocess walk the page links, then kill the subprocess before starting the next page? That'd force a memory release, at least.

Or, don't walk all pages.  Originally wanted this in case docs loaded suddenly onto older session pages.  But from the look of the data so far, when old docs from earlier sessions are suddenly tabled, they appear in the current session, but with their "tabled session" field set to the old session id.  So, probably only need to walk the latest session (after the initial crawl of everything). 

Alternatively, could abandon scrapy and hunt other options.

**Comment by otherchirps** [comment]
So far have found:
- My lazy option of using scraperwiki.sql.save() isn't contributing to the memory usage.  It does save to disk immediately.
- It's not a memory leak by my code itself, no garbage / held references.
- It's not the scrapy framework itself.  Have tried an alternative "lightweight" framework (See: 3b1c7e0942bf133eb8e4a6e1eabee0806cd81d52), but it managed to hit 200MB of usage as well.

It's looking like it may simply be the sheer number of little dictionaries flying around, grabbing pages from the OS, and just being hoarded by the python runtime "just in case" they're needed again (that's actually a feature ... sometimes).

May need to go multiprocess.

eg. 
1. small, master process launches parliament session index scraper child process,
2. Stash the list of parliament session urls to file, child process exits.
3. for each session url, master launches a child worker for that page. then the child exits.

Aiming to cut down the number of items each process spawns.

**Comment by otherchirps** [comment]
One page per child process is no good. 
Implemented this in ba24256118a7722c93250139587b249e92d9ed9f.

Works for most pages, but some pages manage to hit the ceiling on their own. 
eg:

```
[legislative_assembly_tabled_docs] DEBUG: Scraped from <200 http://www.parliament.nsw.gov.au/prod/la/latabdoc.nsf/V3ListBySession?open&key=551>
{'date_tabled': '2014-08-14',
 'laid_by': ...
}
2015-06-10 12:50:06+1000 [legislative_assembly_tabled_docs] INFO: Memory usage: 113008 (kb)
2015-06-10 12:50:06+1000 [legislative_assembly_tabled_docs] INFO: Closing spider (memory_exceeded)
2015-06-10 12:50:06+1000 [legislative_assembly_tabled_docs] ERROR: spider_closed signal received: memory_exceeded
```

The content of this page is a touch over 2MB. 
Could go further with this, (eg. more granular than a page) but this attempt shows that the root cause of the runaway memory usage is still unaddressed.

**Comment by otherchirps** [comment]
As seen [here](http://stackoverflow.com/a/5260513/75188), it might be that the xpath results are giving "smart strings", instead of strings.  This would hold open a reference to the original doc...

http://lxml.de/xpathxslt.html#xpath-return-values

Not the smart string thing.... scrapy explicitly turns those off in its LxmlSel wrappers.  

If each page is around 2MB, or so, this would easily eat up the 20+ MB of memory (assuming the baseline of 80MB is valid -- seems to be).

**Comment by otherchirps** [comment]
Can now be lazy about this.  Morph upped the memory limit: openaustralia/morph#898 :sparkles: :metal:
